api/twitter/tokenProvider.py

```diff
@@ -0,0 +1,137 @@
+                    )
+                )
+                return driver
+            except Exception as exc:
+                print("0 - ERROR -> %s. Crawler could not connect to the database. Retrying in 5s..." % exc)
+                time.sleep(5)
+                continue
+
+    def __get_user_auth(self):
+        """
+        Return old token by updating the reset time
+        and get a free token.
+        """
+        if hasattr(self, 'twitter_api'):
+            self.__update_reset_time()
+        # get new credentials from the db and block them
+        # for 1000min by setting the timestamp
+        user_token = ""
+        user_secret = ""
+        while True:
+            actual_time = math.floor(time.time())
+            distant_time = actual_time + (60 * 1000)
+            query = "MATCH (h:TOKEN) "
+            query += "WHERE NOT EXISTS(h.`%s`) " % self.twitter_route
+            query += "OR h.`%s` < %s " % (self.twitter_route, actual_time)
+            query += "WITH h LIMIT 1 "
+            query += "SET h.`%s`=%s " % (self.twitter_route, distant_time)
+            query += "RETURN h.token as token, "
+            query += "h.secret as secret, "
+            query += "h.user as user"
+            with self.neo4j_driver.session() as db:
+                results = db.run(query)
+                for user in results:
+                    user_token = user["token"]
+                    user_secret = user["secret"]
+            if not user_token:
+                time.sleep(10)
+                print('All token are busy.... waiting.')
+                continue
+            else:
+                break
+        self.user_token = user_token
+        self.api = TwitterAPI(
+            self.app_token,
+            self.app_secret,
+            user_token,
+            user_secret)
+
+    def __update_reset_time(self):
+        # update token's timestamp to real reset time
+        while True:
+            try:
+                response = self.api.request(self.RATE_LIMIT, {})
+                break
+            except Exception as exc:
+                print("13 - ERROR -> %s. ConnectTimeout retrying in 10s..." % exc)
+                time.sleep(10)
+                continue
+        parsed_response = json.loads(response.text)['resources']
+        try:
+            for category in parsed_response.keys():
+                for route in parsed_response[category].keys():
+                    if self.twitter_route in route:
+                        reset_time = parsed_response[category][route]['reset']
+            query = "MATCH (h:TOKEN{token:'%s'}) " % self.user_token
+            query += "SET h.`%s`=%s" % (self.twitter_route, reset_time)
+            self.__run_query(query)
+        except Exception as exc:
+            print("14 - ERROR -> %s. Failed to reset the token's timestamp." % exc)
+            if 'errors' in parsed_response:
+                e_code = parsed_response['errors'][0]['code']
+                if e_code == 89:
```

Comment:
At some point, we should have a concise mapping from twitter error codes to corresponding error message. We could maintain them in a file `constants.py`.